File: p0/p72/MinDistance.java
package leetcode.p0.p72;

public class MinDistance {
    /**
     * DP
     * Time complexity: O(mn)
     * Space complexity: O(mn)
     */
    public static int minDistance1(String word1, String word2) {
        int[][] dp = new int[word1.length() + 1][word2.length() + 1];
        dp[0][0] = 0;
        for (int i = 1; i < dp.length; i++) {
            dp[i][0] = i;
        }
        for (int j = 1; j < dp[0].length; j++) {
            dp[0][j] = j;
        }

        for (int i = 1; i < dp.length; i++) {
            for (int j = 1; j < dp[0].length; j++) {
                if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    dp[i][j] = Math.min(dp[i - 1][j], Math.min(dp[i][j - 1], dp[i - 1][j - 1] - 1)) + 1;
                } else {
                    dp[i][j] = Math.min(dp[i - 1][j], Math.min(dp[i][j - 1], dp[i - 1][j - 1])) + 1;
                }
            }
        }
        return dp[word1.length()][word2.length()];
    }

    /**
     * DP
     * Time complexity: O(mn)
     * Space complexity: O(n)
     */
    public static int minDistance(String word1, String word2) {
        int[] dp = new int[word2.length() + 1];
        dp[0] = 0;
        for (int j = 1; j < dp.length; j++) {
            dp[j] = j;
        }

        for (int i = 1; i < word1.length() + 1; i++) {
            int prev = dp[0];
            dp[0] = i;
            for (int j = 1; j < dp.length; j++) {
                int temp = dp[j];
                if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    dp[j] = Math.min(temp, Math.min(dp[j - 1], prev - 1)) + 1;
                } else {
                    dp[j] = Math.min(temp, Math.min(dp[j - 1], prev)) + 1;
                }
                prev = temp;
            }
        }
        return dp[dp.length - 1];
    }

    public static void main(String[] args) {
        minDistance1("horse", "ros");
        minDistance("horse", "ros");
    }
}

File: p0/p72/min_distance.py
class Solution:
    """
    Feb 19, 2025 14:56
    Time Complexity: O(mn)
    Space Complexity: O(mn)
    """
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)

        if m == 0:
            return n
        if n == 0:
            return m

        dp = []
        for i in range(m):
            dp.append([0] * n)

        if word1[0] == word2[0]:
            dp[0][0] = 0
        else:
            dp[0][0] = 1

        for i in range(1, m):
            if word1[i] == word2[0]:
                dp[i][0] = i
            else:
                dp[i][0] = dp[i - 1][0] + 1

        for i in range(1, n):
            if word1[0] == word2[i]:
                dp[0][i] = i
            else:
                dp[0][i] = dp[0][i - 1] + 1

        for i in range(1, m):
            for j in range(1, n):
                if word1[i] == word2[j]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + 1)

        return dp[-1][-1]